File: updater/src/main/java/updater/identifiers/TriBool.java
package updater.identifiers;

import org.objectweb.asm.Type;
import updater.utility.AbstractIdentifier;
import updater.utility.ClassWrapper;

import java.lang.reflect.Modifier;

public class TriBool extends AbstractIdentifier {
    @Override
    public boolean isMatch(ClassWrapper classNode) {
        return classNode.getInstanceFields().isEmpty() &&
                classNode.getInstanceMethods().isEmpty() &&
                classNode.isOwnerless() &&
                classNode.getCountStaticFieldsOfType(Type.getObjectType(classNode.getName())) == 3 &&
                classNode.getStaticFields().stream().filter(fieldNode -> Type.getType(fieldNode.desc).equals(Type.getObjectType(classNode.getName()))).allMatch(fieldNode -> Modifier.isPublic(fieldNode.access));
    }
}

File: find-names/src/main/java/mapper/identifiers/classes/TriBool.kt
package mapper.identifiers.classes

import mapper.abstractclasses.IdentityMapper
import mapper.predicateutilities.and
import mapper.predicateutilities.predicateOf
import mapper.predicateutilities.type
import mapper.wrappers.ClassWrapper
import java.lang.reflect.Modifier

class TriBool : IdentityMapper.Class() {

    override val predicate = predicateOf<ClassWrapper> { it.instanceFields.isEmpty() }
            .and { it.instanceMethods.isEmpty() }
            .and { it.superType == Any::class.type }
            .and { a -> a.staticFields.count { it.type == a.type } == 3 }
            .and { a -> a.staticFields.filter { it.type == a.type }.all { Modifier.isPublic(it.access) } }
}